### ISSUE TITLE
How to download all subs but only english auto subs ?

### ISSUE THREAD

**Issue by chrizilla** [yt-dlp/yt-dlp#2262]
### Checklist

- [X] I'm asking a question and **not** reporting a bug/feature request
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions including closed ones

### Question

What parameters do I have to use if I want to download ALL subtitles but only English automatic subtitles (instead of the myriad of automatic subtitles in all existing languages on youtube) ?

`--sub-langs` seems to control both `--write-subs` and `--write-auto-subs`
whereas I would need something like `--sub langs all --auto-sub-langs en.*` which doesn't exist, right ?

**Comment by pukkandan** [comment]
Right. You'll have to run yt-dlp twice, once with `--sub-langs all --write-subs` and again with `--sub-langs en --write-auto-subs`

**Comment by chrizilla** [comment]
thanks for the feedback. If you don't mind, I will file a feature request later on.
(because it is easy to forget to run yt-dlp a second time when a user is downloading many files and also when a user has not used yt-dlp for a while)

**Comment by pukkandan** [comment]
I have changed this to a feature request. But most people would expect `--sub-lang` to keep behaving as it currently does. Any implementation must not change that.

There are also some other issues with the current subtitle selection options:
1. If 2 subs have same language code (happens often for generic), you cant select which to download
2. #826 (see #630)

If possible, I want to have a single solution to all these issues. I was thinking of allowing `--format` like syntax for `--sub-format` which should address all of these. But this is a complicated to implement

**Comment by chrizilla** [comment]
> @pukkandan [said](https://github.com/yt-dlp/yt-dlp/issues/2262#issuecomment-1007921078):
If possible, I want to have a single solution to all these issues.

I updated the OP. Maybe this is the kind of "single solution" you are looking for ?
   
  

> @pukkandan [said](https://github.com/yt-dlp/yt-dlp/issues/2262#issuecomment-1007921078):
>     1. If 2 subs have same language code (happens often for generic), you cant select which to download

`--write-all-thumbnails` handles it this way:
```
videotitle.1.jpg
videotitle.2.jpg
videotitle.3.jpg
etc.
```
Couldn't it be handled the same way? So 2 subs have the same language code, both are downloaded like this:
videotitle.1.en.vtt
videotitle.2.en.vtt
   

> @pukkandan [said](https://github.com/yt-dlp/yt-dlp/issues/2262#issuecomment-1007921078):
>     2. [[Feature Request] Add `--embed-auto-subs` #826](https://github.com/yt-dlp/yt-dlp/issues/826) (see [Subtitle file is not deleted after being embedded #630](https://github.com/yt-dlp/yt-dlp/issues/630))

see OP

**Comment by NewUserHa** [comment]
probably using `--sub-langs` to select both 'normal subs' and 'auto-subs' can keep compatibility and make this issue easier to PR, than adding new independent options

**Comment by NewUserHa** [comment]
another simplest way without breaking compatibility is ignoring all auto-translated subtitles (since they can't be translated right at all) unless users explicitly choose them.

**Comment by pukkandan** [comment]
> another simplest way without breaking compatibility is ignoring all auto-translated subtitles (since they can't be translated right at all) unless users explicitly choose them.

see `skip=translated_subs` option in https://github.com/yt-dlp/yt-dlp#youtube. But how does that solve this issue?

**Comment by NewUserHa** [comment]
Cool. don't know that option since hasn't seen that option in the help.

the main issue is with `--write-subs --write-auto-subs --sub-langs user_langs` will always download the unwanted low-quality ones.
with `--write-subs --write-auto-subs --sub-langs all skip=translated_subs` together, I guess users can download all good-quality subtitles as the player of youtube shows(the amount is <3 usually), and just delete the unwanted ones without checking their quality.
if `--write-subs --write-auto-subs --sub-langs users_wanted_langs skip=translated_subs` can work, then the issue probably is solved.

**Comment by NewUserHa** [comment]
In practice, users should always prefer 'auto-generated closed captions + good translated their language subtitles' by default. and the former one can provide like `[music]` tips which could be useful for some users maybe. so it can be even better if it's able to download like both 'en+en(auto-gened CC)'.
don't know if `--write-subs --write-auto-subs --sub-langs 'en, en.*' skip=translated_subs` is able to.

I guess nobody wants auto-translated subtitles because of inaccurate or completely wrong content. so skip=translated_subs could probably be independent with `hls` or `dash` options.

**Comment by HobbyistDev** [comment]
> probably using `--sub-langs` to select both 'normal subs' and 'auto-subs' can keep compatibility and make this issue easier to PR, than adding new independent options

I agree with this. I think it's better if `--sub-langs` can accept auto-sub and original subs. Eg : `--sub-langs all,auto["en"]` which means all originals sub downloaded and only downloaded auto-sub for en. 

I think `--write-subs` and `--write-auto-subs` can be simplified too like `--write-subs auto` as the same as `--write-auto-subs` and `--write-subs` will have default value like `default` or `original`

**Comment by christoph-heinrich** [comment]
Would be great if there was some way of differentiating automatically generated subtitles from manually created ones in the final file.
Maybe by adding a `-auto` suffix to the name or something.

**Comment by ToyKeeper** [comment]
Just another +1 on this.  Would be nice if there was a way to make it automatically do the right thing.  Even better if it did the right thing by default.

... and by "right thing", I mean embedding every subtitle plus the original language's automatic caption(s).

For the automatic captions, it'd also be nice if it matched the timing to what youtube's player displays, where each word appears as it is being said.  That's a separate issue though.

**Comment by scrutinizer11** [comment]
Chipping in: I'd like to set precedence by making sure the "orig" subs get downloaded first, and when they're missing then go for auto-subs.

Is it how the options work? Will the options as set out below, act as I've described?

` --write-subs --sub-langs "(en|ru)-orig,en"  --write-auto-subs`

**Comment by NewUserHa** [comment]
Answer: No, all subs you selected will be downloaded at once.

for users' view, Youtube has 2 types of subs: 1: auto-generated + uploader uploaded; 2: auto-translated.
for yt-dlp and youtube-dl, Youtube has 2(+2) types of subs: 1: uploader uploaded; 2: auto-subs ( auto-generated, auto-translated, automaticle subs(the content is auto-translated as well) ).

so if you want to achieve that, you need to download them in two runs.

---
I think it's better to change to the users' view, then it also can easily filter auto-generated out if the uploader uploaded subs only 
 are wanted with ` --write-subs --sub-langs "*orig, en, ...`, to solve this usual problem/issue.

**Comment by ToyKeeper** [comment]
Quite often, the auto-generated subtitles on youtube are better than the manually-uploaded ones.  It depends on the channel though, and how much effort they put into subtitles.  So it'd be nice if yt-dlp would grab and embed both in a single run.

**Comment by chrizilla** [comment]
@pukkandan : Have you seen the table in my OP ?
_(Just making sure, because I added the table to the OP after you have replied to it.)_
I hope you like it ?
Inside the table there are links to issues/RFEs solvable by the linkified switch if implemented.

**Comment by omidraha** [comment]
I want to download sub title for auto generated `en` language, 

And I tried this command:

```
yt-dlp VIDEO-ID  --skip-download --no-write-subs    --write-auto-sub --sub-langs en --convert-subtitles srt 
```
But instead of downloading auto generation of `en` sub title, it downloading  auto generated of  `Dutch` language.

![image](https://github.com/yt-dlp/yt-dlp/assets/397672/e5184365-95d0-41b0-858a-21f33ae7c709)

![image](https://github.com/yt-dlp/yt-dlp/assets/397672/97658e5f-7177-4831-ab83-5cc8bfdc3153)

**Comment by pukkandan** [comment]
@omidraha The language may not be given as `en`. See `--list-subs`. If you have further questions, open a new issue.

**Comment by omidraha** [comment]
That was.

```
en       English               vtt, ttml, srv3, srv2, srv1, json3
```

```
[info] tk_CsEO5Ht4: Downloading subtitles: en
[info] Available automatic captions for tk_CsEO5Ht4:
Language Name                  Formats
af       Afrikaans             vtt, ttml, srv3, srv2, srv1, json3
ak       Akan                  vtt, ttml, srv3, srv2, srv1, json3
sq       Albanian              vtt, ttml, srv3, srv2, srv1, json3
am       Amharic               vtt, ttml, srv3, srv2, srv1, json3
ar       Arabic                vtt, ttml, srv3, srv2, srv1, json3
hy       Armenian              vtt, ttml, srv3, srv2, srv1, json3
as       Assamese              vtt, ttml, srv3, srv2, srv1, json3
ay       Aymara                vtt, ttml, srv3, srv2, srv1, json3
az       Azerbaijani           vtt, ttml, srv3, srv2, srv1, json3
bn       Bangla                vtt, ttml, srv3, srv2, srv1, json3
eu       Basque                vtt, ttml, srv3, srv2, srv1, json3
be       Belarusian            vtt, ttml, srv3, srv2, srv1, json3
bho      Bhojpuri              vtt, ttml, srv3, srv2, srv1, json3
bs       Bosnian               vtt, ttml, srv3, srv2, srv1, json3
bg       Bulgarian             vtt, ttml, srv3, srv2, srv1, json3
my       Burmese               vtt, ttml, srv3, srv2, srv1, json3
ca       Catalan               vtt, ttml, srv3, srv2, srv1, json3
ceb      Cebuano               vtt, ttml, srv3, srv2, srv1, json3
zh-Hans  Chinese (Simplified)  vtt, ttml, srv3, srv2, srv1, json3
zh-Hant  Chinese (Traditional) vtt, ttml, srv3, srv2, srv1, json3
co       Corsican              vtt, ttml, srv3, srv2, srv1, json3
hr       Croatian              vtt, ttml, srv3, srv2, srv1, json3
cs       Czech                 vtt, ttml, srv3, srv2, srv1, json3
da       Danish                vtt, ttml, srv3, srv2, srv1, json3
dv       Divehi                vtt, ttml, srv3, srv2, srv1, json3
nl-orig  Dutch (Original)      vtt, ttml, srv3, srv2, srv1, json3
nl       Dutch                 vtt, ttml, srv3, srv2, srv1, json3
en       English               vtt, ttml, srv3, srv2, srv1, json3
eo       Esperanto             vtt, ttml, srv3, srv2, srv1, json3
et       Estonian              vtt, ttml, srv3, srv2, srv1, json3
ee       Ewe                   vtt, ttml, srv3, srv2, srv1, json3
fil      Filipino              vtt, ttml, srv3, srv2, srv1, json3
fi       Finnish               vtt, ttml, srv3, srv2, srv1, json3
fr       French                vtt, ttml, srv3, srv2, srv1, json3
gl       Galician              vtt, ttml, srv3, srv2, srv1, json3
lg       Ganda                 vtt, ttml, srv3, srv2, srv1, json3
ka       Georgian              vtt, ttml, srv3, srv2, srv1, json3
de       German                vtt, ttml, srv3, srv2, srv1, json3
el       Greek                 vtt, ttml, srv3, srv2, srv1, json3
gn       Guarani               vtt, ttml, srv3, srv2, srv1, json3
gu       Gujarati              vtt, ttml, srv3, srv2, srv1, json3
ht       Haitian Creole        vtt, ttml, srv3, srv2, srv1, json3
ha       Hausa                 vtt, ttml, srv3, srv2, srv1, json3
haw      Hawaiian              vtt, ttml, srv3, srv2, srv1, json3
iw       Hebrew                vtt, ttml, srv3, srv2, srv1, json3
hi       Hindi                 vtt, ttml, srv3, srv2, srv1, json3
hmn      Hmong                 vtt, ttml, srv3, srv2, srv1, json3
hu       Hungarian             vtt, ttml, srv3, srv2, srv1, json3
is       Icelandic             vtt, ttml, srv3, srv2, srv1, json3
ig       Igbo                  vtt, ttml, srv3, srv2, srv1, json3
id       Indonesian            vtt, ttml, srv3, srv2, srv1, json3
ga       Irish                 vtt, ttml, srv3, srv2, srv1, json3
it       Italian               vtt, ttml, srv3, srv2, srv1, json3
ja       Japanese              vtt, ttml, srv3, srv2, srv1, json3
jv       Javanese              vtt, ttml, srv3, srv2, srv1, json3
kn       Kannada               vtt, ttml, srv3, srv2, srv1, json3
kk       Kazakh                vtt, ttml, srv3, srv2, srv1, json3
km       Khmer                 vtt, ttml, srv3, srv2, srv1, json3
rw       Kinyarwanda           vtt, ttml, srv3, srv2, srv1, json3
ko       Korean                vtt, ttml, srv3, srv2, srv1, json3
kri      Krio                  vtt, ttml, srv3, srv2, srv1, json3
ku       Kurdish               vtt, ttml, srv3, srv2, srv1, json3
ky       Kyrgyz                vtt, ttml, srv3, srv2, srv1, json3
lo       Lao                   vtt, ttml, srv3, srv2, srv1, json3
la       Latin                 vtt, ttml, srv3, srv2, srv1, json3
lv       Latvian               vtt, ttml, srv3, srv2, srv1, json3
ln       Lingala               vtt, ttml, srv3, srv2, srv1, json3
lt       Lithuanian            vtt, ttml, srv3, srv2, srv1, json3
lb       Luxembourgish         vtt, ttml, srv3, srv2, srv1, json3
mk       Macedonian            vtt, ttml, srv3, srv2, srv1, json3
mg       Malagasy              vtt, ttml, srv3, srv2, srv1, json3
ms       Malay                 vtt, ttml, srv3, srv2, srv1, json3
ml       Malayalam             vtt, ttml, srv3, srv2, srv1, json3
mt       Maltese               vtt, ttml, srv3, srv2, srv1, json3
mi       Māori                 vtt, ttml, srv3, srv2, srv1, json3
mr       Marathi               vtt, ttml, srv3, srv2, srv1, json3
mn       Mongolian             vtt, ttml, srv3, srv2, srv1, json3
ne       Nepali                vtt, ttml, srv3, srv2, srv1, json3
nso      Northern Sotho        vtt, ttml, srv3, srv2, srv1, json3
no       Norwegian             vtt, ttml, srv3, srv2, srv1, json3
ny       Nyanja                vtt, ttml, srv3, srv2, srv1, json3
or       Odia                  vtt, ttml, srv3, srv2, srv1, json3
om       Oromo                 vtt, ttml, srv3, srv2, srv1, json3
ps       Pashto                vtt, ttml, srv3, srv2, srv1, json3
fa       Persian               vtt, ttml, srv3, srv2, srv1, json3
pl       Polish                vtt, ttml, srv3, srv2, srv1, json3
pt       Portuguese            vtt, ttml, srv3, srv2, srv1, json3
pa       Punjabi               vtt, ttml, srv3, srv2, srv1, json3
qu       Quechua               vtt, ttml, srv3, srv2, srv1, json3
ro       Romanian              vtt, ttml, srv3, srv2, srv1, json3
ru       Russian               vtt, ttml, srv3, srv2, srv1, json3
sm       Samoan                vtt, ttml, srv3, srv2, srv1, json3
sa       Sanskrit              vtt, ttml, srv3, srv2, srv1, json3
gd       Scottish Gaelic       vtt, ttml, srv3, srv2, srv1, json3
sr       Serbian               vtt, ttml, srv3, srv2, srv1, json3
sn       Shona                 vtt, ttml, srv3, srv2, srv1, json3
sd       Sindhi                vtt, ttml, srv3, srv2, srv1, json3
si       Sinhala               vtt, ttml, srv3, srv2, srv1, json3
sk       Slovak                vtt, ttml, srv3, srv2, srv1, json3
sl       Slovenian             vtt, ttml, srv3, srv2, srv1, json3
so       Somali                vtt, ttml, srv3, srv2, srv1, json3
st       Southern Sotho        vtt, ttml, srv3, srv2, srv1, json3
es       Spanish               vtt, ttml, srv3, srv2, srv1, json3
su       Sundanese             vtt, ttml, srv3, srv2, srv1, json3
sw       Swahili               vtt, ttml, srv3, srv2, srv1, json3
sv       Swedish               vtt, ttml, srv3, srv2, srv1, json3
tg       Tajik                 vtt, ttml, srv3, srv2, srv1, json3
ta       Tamil                 vtt, ttml, srv3, srv2, srv1, json3
tt       Tatar                 vtt, ttml, srv3, srv2, srv1, json3
te       Telugu                vtt, ttml, srv3, srv2, srv1, json3
th       Thai                  vtt, ttml, srv3, srv2, srv1, json3
ti       Tigrinya              vtt, ttml, srv3, srv2, srv1, json3
ts       Tsonga                vtt, ttml, srv3, srv2, srv1, json3
tr       Turkish               vtt, ttml, srv3, srv2, srv1, json3
tk       Turkmen               vtt, ttml, srv3, srv2, srv1, json3
uk       Ukrainian             vtt, ttml, srv3, srv2, srv1, json3
ur       Urdu                  vtt, ttml, srv3, srv2, srv1, json3
ug       Uyghur                vtt, ttml, srv3, srv2, srv1, json3
uz       Uzbek                 vtt, ttml, srv3, srv2, srv1, json3
vi       Vietnamese            vtt, ttml, srv3, srv2, srv1, json3
cy       Welsh                 vtt, ttml, srv3, srv2, srv1, json3
fy       Western Frisian       vtt, ttml, srv3, srv2, srv1, json3
xh       Xhosa                 vtt, ttml, srv3, srv2, srv1, json3
yi       Yiddish               vtt, ttml, srv3, srv2, srv1, json3
yo       Yoruba                vtt, ttml, srv3, srv2, srv1, json3
zu       Zulu                  vtt, ttml, srv3, srv2, srv1, json3
```

**Comment by skyler14** [comment]
Can I receive clarification on what the API for this appears like on the python scripting side? I would like to default to downloading all manually subtitles, then allow them to download ai generated, both, or none (doing that atm). Where is this accessed in the python API?

**Comment by bashonly** [comment]
@skyler14 you can use [`devscripts/cli_to_api.py`](https://github.com/yt-dlp/yt-dlp/blob/master/devscripts/cli_to_api.py) to convert CLI args to API params

e.g. you can run `python devscripts/cli_to_api.py --write-sub --write-auto-sub --sub-lang all,-live_chat` and it will give you:
```py
ydl_opts = {
    'writesubtitles': True,
    'writeautomaticsub': True,
    'subtitleslangs': ['all', '-live_chat'],
}
 ```
 which is how to download all manually uploaded subs and fallback to all automatic captions.
 
 To download both types of subs, you would need to run two separate instances of yt-dlp, but you could extract the video info only once and reuse it in the 2nd invocation:
 
 ```py
 import yt_dlp
 
 URL = 'test:youtube'
 
manual_subs_opts = {
    'writesubtitles': True,
    'subtitleslangs': ['all', '-live_chat'],
}

with yt_dlp.YoutubeDL(manual_subs_opts) as ydl:
    info = ydl.extract_info(URL)

auto_subs_opts = {
    'writeautomaticsub': True,
    'subtitleslangs': ['all', '-live_chat'],
    'skip_download': True,
}

with yt_dlp.YoutubeDL(auto_subs_opts) as ydl:
    ydl.process_ie_result(info)
 ```

**Comment by candrapersada** [comment]
how to use embed subs without leaving the `.vtt subtitle` `--embed-subs --write-subs --write-auto-subs --sub-langs en.*`

**Comment by escape0707** [comment]
Can we have a quick official patch for this problem since there are https://github.com/yt-dlp/yt-dlp/issues/13831 and https://github.com/yt-dlp/yt-dlp/issues/3875 and with these problems combines I can't really use MPV to watch YouTube videos with subtitles without getting too much "false positive" subtitle choices? And it's also a problem when downloading videos.

I really don't need the low quality auto-translate subs, but I do need to download the auto caption sub since that's usually quite useful and usually available.
